File: EX2.2/ex2_2.scala
import scala.util.control.Exception._

object ex2_2 {

/*
 * n=0のとき： 0
 * n=1のとき： 1
 * n=2のとき： 0 + 1 = 1
 * n=3のとき： 1 + 1 = 2
 * n=4のとき： 1 + 2 = 3
 * ・・・
 */

  def isSorted[A] (as: Array[A], ordered: (A,A) => Boolean): Boolean = {

    @annotation.tailrec
    def loop(n: Int): Boolean =
      if (n >= as.length - 1) true
      else if (ordered(as(n), as(n+1))) false
      else loop(n + 1)

    loop(0)
  }

  def main(args: Array[String]): Unit = {
    val testString1 = Array("abc", "def", "dek", "pru")
    val testString2 = Array("abc", "def", "acd", "klm")
    val testInt1 = Array(1, 3, 4, 5)
    val testInt2 = Array(1, 4, 2, 5)

    println("result of " + testString1 + " is " + isSorted(testString1, (x: String, y: String) => x > y)) 
    println("result of " + testString2 + " is " + isSorted(testString2, (x: String, y: String) => x > y)) 
    println("result of " + testInt1 + " is " + isSorted(testInt1, (x: Int, y: Int) => x > y)) 
    println("result of " + testInt2 + " is " + isSorted(testInt2, (x: Int, y: Int) => x > y)) 

    
//    val testString = List(testString1, testString2)
//    testString.foreach {
//      item => 
//        println("result of " + item + " is " + isSorted(item, (x: String, y: String) => x > y)) 
//    }
//    val testInt = List(testInt1, testInt2)
//    testInt.foreach {
//      item => 
//        println("result of " + item + " is " + isSorted(item, (x: Int, y: Int) => x > y))
//    }
  }
}

File: EX2.2/ex2_2.clj
#!/usr/bin/env inlein

; 本当はInt型とかでもcompare使えるけど実験なのであえて>を使うようにしてみる。

'{:dependencies [[org.clojure/clojure "1.8.0"]]}

(ns ex2_2)

(defmulti compAB
  (fn [x y] [(class x) (class y)]))

(defmethod compAB [String String] [x y] 
  (if (< (compare x y) 0) false true))
(defmethod compAB [Number Number] [x y]
  (< x y))

(def isSorted
  (fn [n]
    (loop [cnt 0]
      (cond
        (= cnt ( - (count n) 1)) true
        (compAB (aget n cnt) (aget n (+ cnt 1))) false
        :else (recur (inc cnt))))))

(defn isSortedString [x]
  (if (= x true) "is sorted" "is not sorted"))
(def testString1 (into-array ["abc", "def", "dek", "pru"]))
(def testString2 (into-array ["abc", "def", "acd", "klm"]))
(def testInt1 (into-array [1, 3, 4, 5]))
(def testInt2 (into-array [1, 4, 2, 5]))


(println (seq testString1) (isSortedString (isSorted testString1)))
(println (seq testString2) (isSortedString (isSorted testString2)))
(println (seq testInt1) (isSortedString (isSorted testInt1)))
(println (seq testInt2) (isSortedString (isSorted testInt2)))
(System/exit 0)
